File: matlab/fecg_open/utils/findQRSpeaks.m
function [peakI, peakAmp] = findQRSpeaks(ecg, win)
% findQRSpeaks  Detect local maxima in an ECG signal using a sliding window.
%
% PURPOSE
%   Scans the input signal and identifies samples that are strictly the
%   maximum within a symmetric window of half-width win.  Used as a
%   lightweight QRS candidate detector before SQI-based beat refinement.
%
% INPUTS
%   ecg     - (1 x L or L x 1 real) single-channel ECG signal.
%   win     - (scalar positive integer) half-width of the search window in
%             samples.  A sample at index i is a local maximum if it is the
%             maximum of ecg(i-win : i+win).
%
% OUTPUTS
%   peakI   - (1 x P integer) sample indices of detected local maxima.
%   peakAmp - (1 x P real) ECG amplitude at each detected peak.

    L = length(ecg);
    peakI = zeros(size(ecg));
    for i = (1+win):(L-win)
        [M, I] = max(ecg(i-win:i+win));
        if i == (i-win-1+I)
            peakI(i) = 1;
        end
    end
    peakI = find(peakI > 0);
    peakAmp = ecg(peakI);

end

File: python/fecg_open/utils/findQRSpeaks.py
"""
findQRSpeaks: Local maxima detection via sliding-window maximum.

A sample at position i is a peak if and only if it equals the maximum within
the window [i-win, i+win]. Returns 1-based indices (MATLAB convention).
"""

import numpy as np


def findQRSpeaks(ecg, win):
    """
    Find local maxima in ECG signal using sliding-window method.

    Parameters
    ----------
    ecg : ndarray
        ECG signal (1-D, any orientation)
    win : int
        Half-width of sliding maximum window (samples)

    Returns
    -------
    peakI : ndarray
        Sample indices of detected peaks (1-based, MATLAB convention, column vector)
    peakAmp : ndarray
        ECG amplitudes at detected peak indices (column vector)
    """

    ecg = np.asarray(ecg).flatten()
    L = len(ecg)

    # Pre-allocate boolean array to mark peaks
    is_peak = np.zeros(L, dtype=bool)

    # Sliding-window scan: a sample at position i is a peak if and only if it
    # equals the maximum within the window [i-win, i+win].
    for i in range(win, L - win):
        window_segment = ecg[i - win : i + win + 1]
        I = np.argmax(window_segment)
        if I == win:  # peak is at the centre of the window
            is_peak[i] = True

    peakI = np.where(is_peak)[0] + 1
    peakAmp = ecg[is_peak]  # amplitudes at peak positions

    return peakI.astype(int), peakAmp.astype(float)
